frontend_server/lib/server.dart

```diff
@@ -0,0 +1,207 @@
+
+import 'package:args/args.dart';
+import 'package:front_end/compilation_message.dart';
+
+import 'package:front_end/incremental_kernel_generator.dart';
+import 'package:front_end/compiler_options.dart';
+import 'package:front_end/kernel_generator.dart';
+import 'package:kernel/ast.dart';
+import 'package:kernel/kernel.dart';
+import 'package:kernel/target/flutter.dart';
+import 'package:kernel/target/targets.dart';
+import 'package:usage/uuid/uuid.dart';
+
+ArgParser _argParser = new ArgParser(allowTrailingOptions: true)
+  ..addFlag('train',
+      help: 'Run through sample command line to produce snapshot',
+      negatable: false)
+  ..addFlag('incremental',
+      help: 'Run compiler in incremental mode', defaultsTo: false)
+  ..addOption('sdk-root',
+      help: 'Path to sdk root',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk');
+
+String _usage = '''
+Usage: server [options] [input.dart]
+
+If input dart source code is provided on the command line, then the server 
+compiles it, generates dill file and exits.
+If no input dart source is provided on the command line, server waits for 
+instructions from stdin.
+
+Instructions:
+- compile <input.dart>
+- recompile <boundary-key>
+<path/to/updated/file1.dart>
+<path/to/updated/file2.dart>
+...
+<boundary-key>
+- accept
+- reject
+- quit
+
+Output:
+- result <boundary-key>
+<compiler output>
+<boundary-key> [<output.dill>]
+
+Options:
+${_argParser.usage}
+''';
+
+enum _State {
+  READY_FOR_INSTRUCTION,
+  RECOMPILE_LIST
+}
+
+/// Actions that every compiler should implement.
+abstract class CompilerInterface {
+  /// Compile given Dart program identified by `filename` with given list of
+  /// `options`.
+  Future<Null> compile(String filename, ArgResults options);
+  /// Assuming some Dart program was previously compiled, recompile it again
+  /// taking into account some changed(invalidated) sources.
+  Future<Null> recompileDelta();
+  /// Accept results of previous compilation so that next recompilation cycle
+  /// won't recompile sources that were previously reported as changed.
+  void acceptLastDelta();
+  /// Reject results of previous compilation. Next recompilation cycle will
+  /// recompile sources indicated as changed.
+  void rejectLastDelta();
+  /// This let's compiler know that source file identifed by `uri` was changed.
+  void invalidate(Uri uri);
+}
+
+class _FrontendCompiler implements CompilerInterface {
+  IncrementalKernelGenerator _ikg;
+  String _filename;
+
+  @override
+  Future<Null> compile(String filename, ArgResults options) async {
+    final String boundaryKey = new Uuid().generateV4();
+    print("result $boundaryKey");
+    final Uri sdkRoot = _ensureFolderPath(options['sdk-root']);
+    final CompilerOptions compilerOptions = new CompilerOptions()
+      ..sdkRoot = sdkRoot
+      ..strongMode = false
+      ..target = new FlutterTarget(new TargetFlags())
+      ..onError = (CompilationMessage message) {
+        print(message);
```

Comment:
technically you don't guarantee that this doesn't accidentally match the boundary key, but that seems unlikely so ok.